api/odilo.py

```diff
@@ -0,0 +1,532 @@
+        self.client_secret = pin
+        message = self.refresh_creds(credential)
+
+        if 'OK' == message:
+            return credential
+        else:
+            raise PatronAuthorizationFailedException(message)
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Check out a book on behalf of a patron.
+
+        :param patron: a Patron object for the patron who wants
+        to check out the book.
+
+        :param pin: The patron's alleged password.
+
+        :param licensepool: Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :param internal_format: Represents the patron's desired book format.
+
+        :return: a LoanInfo object.
+        """
+        record_id = licensepool.identifier.identifier
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron, format=internal_format)
+
+        response = self.patron_request(
+            patron, pin, self.CHECKOUT_ENDPOINT.replace("{recordId}", record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        response_json = response.json()
+        if response.status_code == 404:
+            if response_json and response_json['errors'] and len(response_json['errors']) > 0:
+                error = response_json['errors'][0]
+                if error['id'] == 'ERROR_DATA_NOT_FOUND':
+                    raise NotFoundOnRemote('Record %s' % record_id)
+
+                elif error['id'] == 'NoCopiesAvailable':
+                    # Clearly our info is out of date.
+                    # self.update_licensepool(identifier.identifier)
+                    raise NoAvailableCopies()
+                elif error['id'] == 'PatronHasExceededCheckoutLimit':
+                    raise PatronLoanLimitReached()
+                else:
+                    raise CannotLoan(error['id'])
+        else:
+            return self.loan_info_from_odilo_checkout(licensepool.collection, response_json)
+
+    def loan_info_from_odilo_checkout(self, collection, checkout):
+        start_date = self.extract_date(checkout, 'startTime')
+        end_date = self.extract_date(checkout, 'endTime')
+
+        return LoanInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            checkout['id'],
+            start_date,
+            end_date,
+            checkout['downloadUrl']
+        )
+
+    def checkin(self, patron, pin, licensepool):
+        checkout_id = licensepool.identifier.identifier
+        url = self.CHECKIN_ENDPOINT.replace("{checkoutId}", checkout_id) + "?patronId=" + patron
+
+        response = self.patron_request(patron, pin, url, method='POST')
+        if response.status_code == 404:
+            data = response.json()
+            if data and data['errors'] and len(response.json()['errors']) > 0:
+                error = data['errors'][0]
```

Comment:
This whole thing of checking all the different ways the errors might or might not be present can be refactored, as well as the part that turns an error message into a circulation exception.

Reply:
ok, I refactorized where I considered it useful.